app/models/voucher.rb

```diff
@@ -1,15 +1,136 @@
 # frozen_string_literal: false
 
 class Voucher < ApplicationRecord
+  acts_as_paranoid
+
+  include CalculatedAdjustments
+
   belongs_to :enterprise
 
+  has_many :adjustments,
+           as: :originator,
+           class_name: 'Spree::Adjustment',
+           inverse_of: :voucher,
+           dependent: :nullify
+
   validates :code, presence: true, uniqueness: { scope: :enterprise_id }
 
+  before_validation :add_calculator
+
+  def self.adjust!(order)
+    return if order.nil?
+
+    # Find open Voucher Adjustment
+    return if order.vouchers.empty?
+
+    # We only support one voucher per order right now, we could just loop on vouchers
+    adjustment = order.vouchers.first
+
+    # Recalculate value
+    amount = adjustment.originator.compute_amount(order)
+
+    if order.additional_tax_total.positive?
+      handle_tax_excluded_from_price(order, amount)
+    else
+      handle_tax_included_in_price(order, amount)
+    end
```

Comment:
I think it's possible that an order has items including tax and items excluding tax and even items having both if they are subject to two tax categories. I think that we need to find all items for each type of tax calculation and then create adjustments for whichever kind is present. It means that we can't use the existing totals and have to add it up ourselves. Or am I overcomplicating it?

Reply:
I am not sure about that, I always though that it was one or the other. Maybe something to clarify with Lynne/Kirsten.

Reply:
Just to summarise here as well: It's possible to have mixed included and excluded tax on an order and even on single items but we are ignoring that for now and implement that later.